horovod/run/runner.py

```diff
@@ -631,32 +651,74 @@ def _run(args):
             for i in range(args.np):
-                pickled_result = read_data_from_kvstore(driver_ip, run_func_server_port,
-                                                        'runfunc_result', str(i))
-                results[i] = cloudpickle.loads(pickled_result)
+                results[i] = read_data_from_kvstore(driver_ip, run_func_server_port,
+                                                    'runfunc_result', str(i))
             return results
         finally:
             run_func_server.shutdown_server()
     else:
         command = args.command
-        _launch_job(args, remote_host_names, settings, nics, command)
+        _launch_job(args, settings, nics, command)
         return None
 
 
+def _run_elastic(args):
+    # construct host discovery component
+    if args.host_discovery_script:
+        discover_hosts = discovery.HostDiscoveryScript(args.host_discovery_script, args.slots)
+    elif args.hosts:
+        available_hosts, available_slots = parse_hosts_and_slots(args.hosts)
+        discover_hosts = discovery.FixedHosts(set(available_hosts), available_slots)
+    else:
+        raise ValueError('One of --host-discovery-script, --hosts, or --hostnames must be provided')
+
+    # horovodrun has to finish all the checks before this timeout runs out.
+    if args.start_timeout:
+        start_timeout = args.start_timeout
+    else:
+        # Lookup default timeout from the environment variable.
+        start_timeout = int(os.getenv('HOROVOD_START_TIMEOUT', '30'))
+
+    tmout = timeout.Timeout(start_timeout,
+                            message='Timed out waiting for {activity}. Please '
+                                    'check connectivity between servers. You '
+                                    'may need to increase the --start-timeout '
+                                    'parameter if you have too many servers.')
+    settings = elastic_settings.ElasticSettings(discovery=discover_hosts,
+                                                num_proc=args.np,
+                                                min_np=args.min_np or args.np,
+                                                max_np=args.max_np,
+                                                verbose=2 if args.verbose else 0,
+                                                ssh_port=args.ssh_port,
+                                                extra_mpi_args=args.mpi_args,
+                                                key=secret.make_secret_key(),
+                                                timeout=tmout,
+                                                output_filename=args.output_filename,
+                                                run_func_mode=args.run_func is not None,
+                                                nics=args.nics)
+
+    if not gloo_built(verbose=(settings.verbose >= 2)):
+        raise ValueError('Gloo support is required to use elastic training, but has not been built.  Ensure CMake is '
```

Comment:
Will this solution ever work for MPI or is this going to only focus on gloo

Reply:
Elastic Horovod is Gloo only for now. If we are ever making it work with MPI, it will be a separate PR which will change this line.

Reply:
MPI will not work in the sense you cannot reinitialize the world communicator without killing the processes. And by MPI standard, if one MPI process dies, the entire job dies.

Reply:
Yes, there are no plans at this time to explore MPI due to the reasons @nvcastet mentioned.  Elastic mode was actually the primary motivation for adding Gloo as an alternative to MPI :)